### PR TITLE
neutron: lbaasv2: optional haproxy when using f5

### PR DESCRIPTION
**Pull request by matelakat**
When people use F5 they might still want to have haproxy as a
non-default. This change will add a raw parameter
add_haproxy_as_non_default to the neutron barclamp/f5. When user sets
that value to true, HAProxy will be added as a non-default lbaasv2
driver to the neutron_lbaas.conf

### CODE REVIEW

**Comment by matelakat** [comment]
[Configuration of multiple LBaaS v2 providers](https://trello.com/c/T1Yns1bx/106-configuration-of-multiple-lbaas-v2-providers)

**Comment by vuntz** [comment]
So if we only consider this branch, this approach would work for me. But:

- I would make the attribute not required in the schema (helps to not diverge from upstream)
- I would drop the migration
- I would possibly just make the attribute (in the cookbook) default to true (more debatable as this would definitely cause restart of neutron by default; on the other hand, I guess this would be wanted anyhow?)

I'm slightly concerned that we would not forward port this patch to upstream because it's slightly too hacky. But I think addressing the need is more important than this concern in the short term; we'll find a way to upstream this properly later on.

**Comment by matelakat** [comment]
> So if we only consider this branch, this approach would work for me. But:
> 
> I would make the attribute not required in the schema (helps to not diverge from upstream)
> I would drop the migration
> I would possibly just make the attribute (in the cookbook) default to true (more debatable as this would definitely cause restart of neutron by default; on the other hand, I guess this would be wanted anyhow?)
> I'm slightly concerned that we would not forward port this patch to upstream because it's slightly too hacky. But I think addressing the need is more important than this concern in the short term; we'll find a way to upstream this properly later on.

Regarding to moving this to upstream, that is not the scope of this change. If we at any point have the same functionality implemented upstream in a different way, we can update this branch to be aligned with that solution / provide a straight upgrade path from this codebase to upstream.

I think leaving out the migration would just make it easier for this change to disappear, and would make this change really a hack.

Regarding to defaulting to having haproxy as a non-default driver, I am thinking that it would mean that only the deployment of this change would result in a configuration file change, which is not expected. If people install this patch as it is, no change will happen to the cloud (configuration files). When they change a value in the proposal, that's when the modifications kick off.

I'd like to leave it as it is and not address your concerns. Hopefully the explanations make it clear why I prefer this way. On the other hand if you still think we need to make those changes, please give it a -1 and I will address those immediately.

**Review by vuntz — APPROVED** [review]
Fine with me if @tpatzig and @gonzolino fully understand that this breaks an upgrade to a newer version of the product at the moment.

**Comment by matelakat** [comment]
This patch is not working as the original code is expecting that F5 and haproxy are mutually exclusive. More code changes needed here.

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb [hunk 1]
```diff
@@ -15,6 +15,10 @@
 
 use_l3_agent = (node[:neutron][:networking_plugin] != "vmware")
 use_lbaas_agent = node[:neutron][:use_lbaas]
+use_haproxy_with_lbaasv2 = node[:neutron][:use_lbaasv2] &&
+  [nil, "", "haproxy"].include?(node[:neutron][:lbaasv2_driver])
+use_lbaasv2_with_f5 = node[:neutron][:use_lbaasv2] && node[:neutron][:lbaasv2_driver] == "f5"
+use_haproxy_with_lbaasv2 |= use_lbaasv2_with_f5 && node[:neutron][:f5][:add_haproxy_as_non_default]
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/network_agents_ha.rb]: Should be `||=` not `|=`

File: chef/cookbooks/neutron/recipes/network_agents.rb [hunk 2]
```diff
@@ -19,19 +19,24 @@
 package node[:neutron][:platform][:dhcp_agent_pkg]
 package node[:neutron][:platform][:metering_agent_pkg]
 
-if node[:neutron][:use_lbaas]
+use_lbaasv1_or_lbaasv2_with_haproxy = node[:neutron][:use_lbaas] &&
+  (!node[:neutron][:use_lbaasv2] || [nil, "", "haproxy"].include?(node[:neutron][:lbaasv2_driver]))
+use_lbaasv2_with_f5 = node[:neutron][:use_lbaas] && node[:neutron][:use_lbaasv2] &&
+  node[:neutron][:lbaasv2_driver] == "f5"
+use_lbaasv1_or_lbaasv2_with_haproxy |= use_lbaasv2_with_f5 && node[:neutron][:f5][:add_haproxy_as_non_default]
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/network_agents.rb]: Should be `||=` not `|=`

**Review by vuntz — APPROVED** [review]

**Review by gonzolino — APPROVED** [review]
Tested in mkcloud, worked flawlessly!

**Pull request merged**